frame/assets/src/functions.rs

```diff
@@ -318,26 +331,46 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 				balance: Zero::zero(),
 				is_frozen: false,
 				reason,
+				depositor: Some(depositor.clone()),
 				extra: T::Extra::default(),
 			},
 		);
 		Ok(())
 	}
 
 	/// Returns a deposit, destroying an asset-account.
-	pub(super) fn do_refund(id: T::AssetId, who: T::AccountId, allow_burn: bool) -> DispatchResult {
+	pub(super) fn do_refund(
+		caller: &T::AccountId,
+		id: T::AssetId,
+		who: &T::AccountId,
+		allow_burn: bool,
+	) -> DispatchResult {
 		let mut account = Account::<T, I>::get(id, &who).ok_or(Error::<T, I>::NoDeposit)?;
 		let deposit = account.reason.take_deposit().ok_or(Error::<T, I>::NoDeposit)?;
 		let mut details = Asset::<T, I>::get(&id).ok_or(Error::<T, I>::Unknown)?;
 		ensure!(details.status == AssetStatus::Live, Error::<T, I>::AssetNotLive);
 		ensure!(account.balance.is_zero() || allow_burn, Error::<T, I>::WouldBurn);
 		ensure!(!account.is_frozen, Error::<T, I>::Frozen);
 
-		T::Currency::unreserve(&who, deposit);
+		let mut refund_to: T::AccountId = who.clone();
+		if let Some(depositor) = account.depositor.take() {
+			// Only the account owner (`who`), the depositor, or the asset admin can remove an
+			// account. That is, a depositor cannot force another's account to exist with zero
+			// balance.
+			ensure!(
+				caller == &depositor || caller == who || caller == &details.admin,
+				Error::<T, I>::NoPermission
+			);
+			refund_to = depositor;
+		}
+		T::Currency::unreserve(&refund_to, deposit);
+		account.depositor = None;
 
 		if let Remove = Self::dead_account(&who, &mut details, &account.reason, false) {
 			Account::<T, I>::remove(id, &who);
 		} else {
+			// deposit has been refunded, need to update `Account`
+			Account::<T, I>::insert(id, &who, account);
```

Comment:
I still think it would be good to test also the case where the account does not get removed after `refund()` and `refund_other()`.

Reply:
> I still think it would be good to test also the case where the account does not get removed after `refund()` and `refund_other()`.

I overlooked the `take_deposit()` that overrides the `reason`. It will never reach the `debug_assert!`